nntrainer/src/network_graph.cpp

```diff
@@ -0,0 +1,364 @@
+    std::shared_ptr<Layer> layer =
+      nntrainer::createLayer(LayerType::LAYER_CONCAT);
+    ensureName(layer, current.getName());
+    layer->num_inputs = current.num_inputs;
+    layer->input_layers.clear();
+    for (unsigned int i = 0; i < current.input_layers.size(); ++i)
+      layer->input_layers.push_back(current.input_layers[i]);
+
+    current.num_inputs = 1;
+    current.input_layers.clear();
+    current.input_layers.push_back(layer->getName());
+    addLayerNode(layer);
+  }
+  return status;
+}
+
+int NetworkGraph::realizeFlattenType(
+  Layer &current, std::vector<std::shared_ptr<Layer>> layers) {
+  if (num_node == 0) {
+    ml_loge("layer is empty");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (current.getType() == LayerType::LAYER_FLATTEN) {
+    ml_loge(
+      "It is not allowed to realize flatten layer, possibly flatten layer is "
+      "added right after flatten");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  std::shared_ptr<Layer> layer =
+    nntrainer::createLayer(LayerType::LAYER_FLATTEN);
+
+  ensureName(layer, current.getName());
+
+  layer->num_inputs = 1;
+  layer->input_layers.clear();
+  layer->input_layers.push_back(current.getName());
+
+  addLayerNode(layer);
+
+  for (unsigned int i = 0; i < layers.size(); ++i) {
+    for (unsigned int j = 0; j < layers[i]->input_layers.size(); ++j) {
+      if (layers[i]->input_layers[j] == current.getName()) {
+        layers[i]->input_layers[j] = layer->getName();
+      }
+    }
+  }
+
+  return ML_ERROR_NONE;
+}
+
+int NetworkGraph::realizeActivationType(
+  Layer &current, std::vector<std::shared_ptr<Layer>> layers) {
+  ActivationType act = current.getActivationType();
+
+  if (act == ActivationType::ACT_NONE) {
+    /// ActivationType::ACT_NONE does not need realization
+    return ML_ERROR_NONE;
+  }
+
+  if (num_node == 0) {
+    ml_loge("layer is empty");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (current.getType() == LayerType::LAYER_ACTIVATION) {
+    ml_loge("It is not allowed to realize ativation layer, possibly layer is "
+            "added right after activation");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  if (act == ActivationType::ACT_UNKNOWN) {
+    ml_loge("cannot realize unknown activation type");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+
+  std::shared_ptr<Layer> layer =
+    nntrainer::createLayer(LayerType::LAYER_ACTIVATION);
+
+  ensureName(layer, current.getName());
+  layer->setActivation(act);
+  layer->num_inputs = 1;
+  layer->input_layers.clear();
+  layer->input_layers.push_back(current.getName());
+
+  addLayerNode(layer);
+
+  for (unsigned int i = 0; i < layers.size(); ++i) {
+    for (unsigned int j = 0; j < layers[i]->input_layers.size(); ++j) {
+      if (layers[i]->input_layers[j] == current.getName()) {
+        layers[i]->input_layers[j] = layer->getName();
+      }
+    }
+  }
+
+  return ML_ERROR_NONE;
+}
+
+int NetworkGraph::addLossLayer(const LossType loss_type) {
```

Comment:
Maybe, we have to deal with multiple outputs and heterogeneous loss type.

Reply:
yes. multiple output will come in PR #714